### PR TITLE
Adding test loaders

### PR DESCRIPTION
**Pull request by Louis-Dupont**
Does not include loading test loaders from recipe or custom test_metrics for simplicity (we can add in following PRs)

---
Example I added on how to use
```python
trainer.train(
    model=model,
    training_params=training_params,
    train_loader=cifar10_train(),
    valid_loader=cifar10_val(),
    test_loaders={
        "Copy of train": cifar10_val(),
        "Copy of valid": cifar10_val(),
    },  # For the sake of the example, we are testing on the training and validation set.
)
```
```
===========================================================
Train epoch 3: 100%|██████████| 196/196 [00:07<00:00, 27.34it/s, Accuracy=0.547,
Validating epoch 3: 100%|██████████| 20/20 [00:01<00:00, 13.26it/s]
Testing dataset="Copy of train": epoch 3: 100%|██████████| 20/20 [00:01<00:00, 1
Testing dataset="Copy of valid": epoch 3: 100%|██████████| 20/20 [00:01<00:00, 1
[2023-07-04 15:20:33] INFO - base_sg_logger.py - Checkpoint saved in /home/louis.dupont/PycharmProjects/super-gradients/checkpoints/demo-train-with-test/ckpt_best.pth
[2023-07-04 15:20:33] INFO - sg_trainer.py - Best checkpoint overriden: validation Accuracy: 0.6025000214576721
===========================================================
SUMMARY OF EPOCH 3
├── Test
│   ├── Copy of train:accuracy = 0.6025
│   │   ├── Best until now = 0.4972 (↗ 0.1053)
│   │   └── Epoch N-1      = 0.4972 (↗ 0.1053)
│   ├── Copy of train:labelsmoothingcrossentropyloss = 1.1426
│   │   ├── Best until now = 1.3962 (↘ -0.2536)
│   │   └── Epoch N-1      = 1.3962 (↘ -0.2536)
│   ├── Copy of train:top5 = 0.9576
│   │   ├── Best until now = 0.9363 (↗ 0.0213)
│   │   └── Epoch N-1      = 0.9363 (↗ 0.0213)
│   ├── Copy of valid:accuracy = 0.6025
│   │   ├── Best until now = 0.4972 (↗ 0.1053)
│   │   └── Epoch N-1      = 0.4972 (↗ 0.1053)
│   ├── Copy of valid:labelsmoothingcrossentropyloss = 1.1426
│   │   ├── Best until now = 1.3962 (↘ -0.2536)
│   │   └── Epoch N-1      = 1.3962 (↘ -0.2536)
│   └── Copy of valid:top5 = 0.9576
│       ├── Best until now = 0.9363 (↗ 0.0213)
│       └── Epoch N-1      = 0.9363 (↗ 0.0213)
├── Training
│   ├── Accuracy = 0.5466
│   │   ├── Best until now = 0.4826 (↗ 0.064)
│   │   └── Epoch N-1      = 0.4826 (↗ 0.064)
│   ├── Labelsmoothingcrossentropyloss = 1.2578
│   │   ├── Best until now = 1.4573 (↘ -0.1995)
│   │   └── Epoch N-1      = 1.4573 (↘ -0.1995)
│   └── Top5 = 0.9465
│       ├── Best until now = 0.9231 (↗ 0.0234)
│       └── Epoch N-1      = 0.9231 (↗ 0.0234)
└── Validation
    ├── Accuracy = 0.6025
    │   ├── Best until now = 0.4972 (↗ 0.1053)
    │   └── Epoch N-1      = 0.4972 (↗ 0.1053)
    ├── Labelsmoothingcrossentropyloss = 1.1426
    │   ├── Best until now = 1.3962 (↘ -0.2536)
    │   └── Epoch N-1      = 1.3962 (↘ -0.2536)
    └── Top5 = 0.9576
        ├── Best until now = 0.9363 (↗ 0.0213)
        └── Epoch N-1      = 0.9363 (↗ 0.0213)
```

Wandb (with other dataset names)

image

### CODE REVIEW

File: src/super_gradients/training/sg_trainer/sg_trainer.py [hunk 1]
```diff
@@ -1070,6 +1092,7 @@ def forward(self, inputs, targets):
         # METRICS
         self._set_train_metrics(train_metrics_list=self.training_params.train_metrics_list)
         self._set_valid_metrics(valid_metrics_list=self.training_params.valid_metrics_list)
+        self.test_metrics = self.valid_metrics.clone()
```
Comment (Louis-Dupont) [inline src/super_gradients/training/sg_trainer/sg_trainer.py:1095]: Until we support custom test metrics, let me know if you want to do differently

File: src/super_gradients/training/sg_trainer/sg_trainer.py [hunk 2]
```diff
@@ -1710,13 +1759,21 @@ def _get_hyper_param_config(self):
...
+        inf_time: float,
+        epoch: int,
+        context: PhaseContext,
     ):
         """Run the various logging operations, e.g.: log file, Tensorboard, save checkpoint etc."""
         result_dict = {
             "Inference Time": inf_time,
             **{f"Train_{k}": v for k, v in train_metrics_dict.items()},
             **{f"Valid_{k}": v for k, v in validation_results_dict.items()},
+            **{f"Test_{k.replace(' ', '_')}": v for k, v in test_metrics_dict.items()},
```
Comment (Louis-Dupont) [inline src/super_gradients/training/sg_trainer/sg_trainer.py:1776]: Question: Are we ok with whitespace in the names of logged values (wandb, tb, ...) ?
And what about the prints, are we ok with having `dataset_name` including whitespace when printing each metric result in format `{dataset_name}:{metric_name}` ?

File: src/super_gradients/training/sg_trainer/sg_trainer.py (hunk 1 above)
Comment (BloodAxe) [reply src/super_gradients/training/sg_trainer/sg_trainer.py:1084]: I'm not sure whether valid_metrics is MetricsCollection or generic list. 
So there is a chance that .clone() is doing a shallow copy which may not be what we are expecting.
Maybe copy.deepcopy() just to be safe?

File: src/super_gradients/training/sg_trainer/sg_trainer.py (hunk 2 above)
Comment (BloodAxe) [reply src/super_gradients/training/sg_trainer/sg_trainer.py:1783]: I personally againts doing any sanitization of strings inside the SG trainer.
First and foremost: Why? 
If some logger does not support whitespaces or `@` symbol (All in all this is wierd, but ok) - the concrete logger class should take care of that. 
That's my 5 cents.

File: src/super_gradients/training/sg_trainer/sg_trainer.py (hunk 1 above)
Comment (Louis-Dupont) [reply src/super_gradients/training/sg_trainer/sg_trainer.py:1091]: It's always a `MetricCollection` but it's not super explicit because `_set_train_metrics` does it inplace.

File: src/super_gradients/training/utils/sg_trainer_utils.py [hunk 3]
```diff
@@ -240,6 +246,7 @@ def _generate_tree(value_name: str, monitored_value: MonitoredValue) -> Tree:
             tree.create_node(tag=f"Best until now = {best:6} ({diff_with_best_colored:8})", identifier=f"1_best_{root_id}", parent=root_id)
         return tree
 
+    print("===========================================================")
```
Comment (BloodAxe) [inline src/super_gradients/training/utils/sg_trainer_utils.py:249]: Debug leftover?

File: src/super_gradients/training/sg_trainer/sg_trainer.py (hunk 1 above)
Comment (Louis-Dupont) [reply src/super_gradients/training/sg_trainer/sg_trainer.py:1092]: Added typing here: https://github.com/Deci-AI/super-gradients/pull/1241/commits/2d51ab6cb000c0dc15728e1aeaeb90dc2840ed16

File: src/super_gradients/training/utils/sg_trainer_utils.py [hunk 4]
```diff
@@ -190,16 +190,22 @@ def update_monitored_values_dict(monitored_values_dict: Dict[str, MonitoredValue
...
     return monitored_values_dict
 
 
 def display_epoch_summary(
-    epoch: int, n_digits: int, train_monitored_values: Dict[str, MonitoredValue], valid_monitored_values: Dict[str, MonitoredValue]
+    epoch: int,
+    n_digits: int,
+    train_monitored_values: Dict[str, MonitoredValue],
+    valid_monitored_values: Dict[str, MonitoredValue],
+    test_monitored_values: Dict[str, MonitoredValue],
```
Comment (BloodAxe) [inline src/super_gradients/training/utils/sg_trainer_utils.py:208]: I feel uneasy about duplicated code. What if we play a little bit future-proof here?
Let's say we add a single parameter: `loader_metrics: Dict[str, Dict[str, MonitoredValue]]` instead of `train_monitored_values, valid_monitored_values, test_monitored_values`.

The call to the method would obviously change, but not that significantly:
```
display_epoch_summary(
  loader_metrics=dict(
    train=train_monitored_values,
    valid=valid_monitored_values,
    test=test_monitored_values
  )
}
```

However internal implementation would be simplified since we would get rid of duplicated code that generates the summary tree for each loader. Instead we could simply iterate over the collection.
If has another advantage that if we will get to the multiple validation loaders - this function will be already ready for this change.
WDYT?

Comment (Louis-Dupont) [reply src/super_gradients/training/utils/sg_trainer_utils.py:208]: Well, I see two different things and am not sure which one you are thinking about:
1. Grouping all  the "splits" monitored values into a single dict: `Dict[str, Dict[str, MonitoredValue]]`
```python
loader_metrics=dict(
    train=train_monitored_values,
    valid=valid_monitored_values,
    test=test_monitored_values
  )
```
2. Allowing each split to support nested monitored values (1 per loader): `Dict[str, Dict[str, MonitoredValue]]`
```python
valid_monitored_values = {
    "dataset_1": dataset1_monitored_values, ...
    "dataset_2": dataset2_monitored_values, ...
}
```
Which one are you referring to?
I see less value in the 1. because we always work with train/test/valid, so I don't see a lot of value into abstracting the split away. I think that the solution would rather be to create a single class MonitoredValues (or a better name) that would handle train/valid/test and take out the code from the Trainer.
What do you think?

Comment (BloodAxe) [reply src/super_gradients/training/utils/sg_trainer_utils.py:208]: I mean the first option)

File: tests/unit_tests/train_after_test_test.py [hunk 5]
```diff
@@ -61,7 +67,13 @@ def test_call_train_after_test_with_loss(self):
             "greater_metric_to_watch_is_better": True,
         }
         trainer.test(model=model, test_metrics_list=[Accuracy()], test_loader=dataloader, loss=torch.nn.CrossEntropyLoss())
-        trainer.train(model=model, training_params=train_params, train_loader=dataloader, valid_loader=dataloader)
+        trainer.train(
```
Comment (shaydeci) [inline tests/unit_tests/train_after_test_test.py:70]: Lets seperate this to a different test case.

File: src/super_gradients/training/sg_trainer/sg_trainer.py [hunk 6]
```diff
@@ -1866,7 +1950,11 @@ def evaluate(
 
             if not silent_mode:
                 # PRINT TITLES
-                pbar_start_msg = f"Validation epoch {epoch}" if evaluation_type == EvaluationType.VALIDATION else "Test"
+                pbar_start_msg = "Validating" if evaluation_type == EvaluationType.VALIDATION else "Testing"
```
Comment (shaydeci) [inline src/super_gradients/training/sg_trainer/sg_trainer.py:1953]: Why did you change the PBar message ?

File: src/super_gradients/training/sg_trainer/sg_trainer.py [hunk 7]
```diff
@@ -1318,17 +1339,39 @@ def forward(self, inputs, targets):
...
+                        monitored_values_dict=self.valid_monitored_values,
+                        new_values_dict=valid_metrics_dict,
+                    )  # TODO: Move this logic inside a MonitoredValues class
                     # Phase.VALIDATION_EPOCH_END
                     # RUN PHASE CALLBACKS
                     context.update_context(metrics_dict=valid_metrics_dict)
                     self.phase_callback_handler.on_validation_loader_end(context)
 
+                test_metrics_dict = {}
+                if (epoch + 1) % self.run_validation_freq == 0:
```
Comment (shaydeci) [inline src/super_gradients/training/sg_trainer/sg_trainer.py:1359]: If you decided that we also use run_validation_freq attribute to when we perform test you need to state this in the docs please.

**Review by shaydeci — CHANGES_REQUESTED** [review]
Im not sure I quite follow how the monitoring is going to be for multiple test datasets in SGLogger, and there seems to be a lack of documentation for this.

File: src/super_gradients/training/sg_trainer/sg_trainer.py (hunk 6 above)
Comment (Louis-Dupont) [reply src/super_gradients/training/sg_trainer/sg_trainer.py:1953]: It's to add the possibility of include `dataset_name`, otherwise we would get `Testing epoch X` for every test dataset, and I thought it would be nicer to show the dataset name as well.
Eventually, it just means that `Validation epoch X` will be replaced with `Validation epoch X`, and for test datasets, it will be `Testing dataset="dataset-name": epoch X` (like in the description of the PR)

File: src/super_gradients/training/sg_trainer/sg_trainer.py (hunk 7 above)
Comment (Louis-Dupont) [reply src/super_gradients/training/sg_trainer/sg_trainer.py]: I added a new param similar to validation: https://github.com/Deci-AI/super-gradients/pull/1241/commits/ab9f715e1dac36c3b10ad5bcc47a12669bd71d7b

File: src/super_gradients/training/utils/sg_trainer_utils.py (hunk 4 above)
Comment (Louis-Dupont) [reply src/super_gradients/training/utils/sg_trainer_utils.py]: Like this? https://github.com/Deci-AI/super-gradients/pull/1241/commits/1d890dd8fe92d8d7643b2e9c155c51a9d89d37dc

File: src/super_gradients/training/sg_trainer/sg_trainer.py (hunk 7 above)
Comment (Louis-Dupont) [reply src/super_gradients/training/sg_trainer/sg_trainer.py]: (I changed the value to 1 here [7321c0c](https://github.com/Deci-AI/super-gradients/pull/1241/commits/7321c0ce0c6a7dc94ff2e67a86168514aed23f8f), that was a mistake)

**Review by shaydeci — APPROVED** [review]
LGTM on my end.
@BloodAxe Has another unresolved comment.

File: src/super_gradients/training/utils/sg_trainer_utils.py [hunk 8]
```diff
@@ -190,28 +190,26 @@ def update_monitored_values_dict(monitored_values_dict: Dict[str, MonitoredValue
...
 
     :param epoch: the number of epoch.
     :param n_digits: number of digits to display on screen for float values
-    :param train_monitored_values: mapping of loss/metric with their stats that will be displayed
-    :param valid_monitored_values: mapping of loss/metric with their stats that will be displayed
-    :return:
+    :param monitored_values_dict: Dict of Dict. The first one represents the splut, and the second one a loss/metric.
     """
 
     def _format_to_str(val: float) -> str:
```
Comment (BloodAxe) [inline src/super_gradients/training/utils/sg_trainer_utils.py:211]: Maybe change `val: float` to `val: Optional[float]` since we now allow val to be None

File: src/super_gradients/training/pre_launch_callbacks/pre_launch_callbacks.py [hunk 9]
```diff
@@ -109,6 +109,7 @@ def __call__(self, cfg: DictConfig) -> DictConfig:
         tmp_cfg.training_hyperparamsbatch_accumulate = 1
         tmp_cfg.training_hyperparamsmax_train_batches = self.num_forward_passes
         tmp_cfg.training_hyperparamsrun_validation_freq = 2
+        tmp_cfg.training_hyperparamsrun_test_freq = 2
```
Comment (BloodAxe) [inline src/super_gradients/training/pre_launch_callbacks/pre_launch_callbacks.py:112]: I feel we have a typo error in this whole section:
`tmp_cfg.training_hyperparamsrun_test_freq = 2`

There is a missing dot after `tmp_cfg.training_hyperparams`:
`tmp_cfg.training_hyperparams.run_test_freq = 2`

Comment (BloodAxe) [reply src/super_gradients/training/pre_launch_callbacks/pre_launch_callbacks.py:112]: @shaydeci how that is supposed to be working in principle if we have that many typos?

Comment (shaydeci) [reply src/super_gradients/training/pre_launch_callbacks/pre_launch_callbacks.py:112]: Completely missed that one.
Please fox @Louis-Dupont

Comment (Louis-Dupont) [reply src/super_gradients/training/pre_launch_callbacks/pre_launch_callbacks.py]: Fixed to `tmp_cfg.training_hyperparams.run_test_freq = 2`

**Review by shaydeci — APPROVED** [review]
